### ISSUE TITLE
CPU memory leak when training with siglip loss

### ISSUE THREAD

**Issue by estherxue** [mlfoundations/open_clip#849]
Hi, I tried to train with siglip loss on a large dataset and I found that during training (not evaluation), CPU memory usage kept increasing. The program was finally killed by the system. The data loading process is nothing special, similar to that of what csv dataset does. Has anyone encountered a similar problem?

**Comment by rwightman** [comment]
@estherxue does it behave differently than with normal CLIP (infonce) loss on the exact setup?

**Comment by estherxue** [comment]
> @estherxue does it behave differently than with normal CLIP (infonce) loss on the exact setup?

Below is the running script for siglip:
`torchrun --nproc_per_node 1 \
         --nnodes $WORLD_SIZE \
         --node_rank $RANK \
         --master_addr $MASTER_ADDR \
         --master_port $MASTER_PORT \
                  -m training.main \
    --train-data '' \
    --val-data '' \
    --dataset-type hugface \
    --batch-size 512 \
    --precision amp \
    --csv-img-key image_hash \
    --csv-caption-key caption \
    --local-loss \
    --gather-with-grad \
    --logs /home/work/data_mm_pretrain/models/siglip_b16_60m_large_bs_no_wd/ \
    --name large_bs \
    --workers 12 \
    --epochs 10 \
    --model ViT-B-16-SigLIP \
    --pretrained webli \
    --warmup 0 \
    --beta2 0.95 \
    --lr 5e-5 \
    --wd 0. \
    --torchcompile \
    --siglip`

Below is the running script for clip:
`torchrun --nproc_per_node 1 \
         --nnodes $WORLD_SIZE \
         --node_rank $RANK \
         --master_addr $MASTER_ADDR \
         --master_port $MASTER_PORT \
         -m training.main \
        --train-data '' \
    --val-data '' \
    --dataset-type hugface \
    --batch-size 768 \
    --precision amp \
    --csv-img-key image_hash \
    --csv-caption-key caption \
    --local-loss \
    --gather-with-grad \
    --logs /home/work/data_mm_pretrain/models/clip_b32_id_2.5m_baseline/ \
    --name large_bs \
    --workers 12 \
    --epochs 4 \
    --model ViT-B-32-quickgelu \
    --pretrained openai \
    --use-thumbnail \
    --warmup 0 \
    --lr 5e-5 \
    --wd 0. \
    --torchcompile`
The only difference that I can notice is the memory usage on CPU.

**Comment by estherxue** [comment]
I found that with CLIP loss the memory leak also exists.
<img width="807" alt="截屏2024-04-12 上午6 38 55" src="https://github.com/mlfoundations/open_clip/assets/17434792/05601545-fd95-42c1-8bc0-3cb4187a8c65">

**Comment by miguelalba96** [comment]
did you tried training with standard clip loss?

**Comment by darkasevgen** [comment]
Hi, are there any updates?

**Comment by estherxue** [comment]
> did you tried training with standard clip loss?

I tried training with standard clip loss. I was wrong. Standard clip loss also has memory leak problem.

**Comment by estherxue** [comment]
> Hi, are there any updates?

I finally solved this problem by only training with limited batch number for each epoch as after the code finishes training for one epoch, the memory usage goes down.

**Comment by estherxue** [comment]
It seems that this has nothing to do with the loss. The memory leak exists when doing evaluation.

**Comment by miguelalba96** [comment]
I did some research on CPU memory leaks, and people say most of the time memory leaks appear when tensors are accumulated without being detached (as they carry with them the entire computational graph) or data loader issues such as copy-on-access (storing in the dataset definition naive python objects which reference count gets increased when being accessed by multiple processes (dataloader workers))

These resources might help you to debug:
- https://github.com/pytorch/pytorch/issues/13246
- https://ppwwyyxx.com/blog/2022/Demystify-RAM-Usage-in-Multiprocess-DataLoader/#Motivation-for-In-RAM-Data

If you find something let me know, I am experiencing RAM memory leaks too when fine-tuning CLIP using LoRA and standard distributed clip loss mentioned implemented in this repo, but I use torch ligthing Fabric launcher and mosaic ml streaming dataset instead of WebDataset

**Comment by abhishek0093** [comment]
Hi, 
I’m trying to distill L-14 model into a B-16. Each model independently is running fine, but when using distilled functionality I’m also getting OOM on cpu after 30% train on the 1st epoch whereas on gpu memory is still available(100% utilisation, 11G/16G used per gpu). I’m also using databricks . The memory consumption is high, but I'm not sure why it is the case. Is it because of large number of workers? I used same number of workers when training models individually(no distillation). This is my training script : 
`%sh 
torchrun --nproc_per_node 8 -m training.main \
  --batch-size 32 \
  --precision amp \
  --name $OPENCLIP_RUN_NAME \
  --workers 64 \
  --report-to tensorboard \
  --save-frequency 1 \
  --dataset-type webdataset \
  --train-data "/root/img_dataset/{0000000..0002100}.tar" \
  --val-data "/root/img_dataset/{0000000..0000001}.tar" \
  --train-num-samples 2100000 \
  --lr 1.0e-6 \
  --epochs 8 \
  --model ViT-B-16 \
  --epochs-cooldown 0 \
  --beta1 0.9 \
  --beta2 0.98 \
  --eps 1.0e-6 \
  --wd 0.2 \
  --pretrained datacomp_xl_s13b_b90k \
  --accum-freq 1 \
  --distill-model ViT-L-14  \
  --distill-pretrained datacomp_xl_s13b_b90k`

And this is my memory consumption stats : 
<img width="450" alt="Screenshot 2024-05-02 at 11 21 18 AM" src="https://github.com/mlfoundations/open_clip/assets/111701999/d46d9375-e826-4988-9a3a-fa24fe979d38">

**Comment by rwightman** [comment]
we've done a lot of large scale training, long durations, big datasets and never found any noteworthy issues with dataloader memory leaks and the webdataset code. We don't use csv datasets though so possibly an issue there.

There is significant memory churn when you're plowing through really large datasets. Some allocators have issues with fragmentation over time. I usually patch the allocator to use tcmalloc. `LD_PRELOAD=/lib/<system dependent>/libtcmalloc.so.4` ... `apt get install google-perftools` to get the lib.

Should point out that normal 'validation' is VERY memory intensive if you have a lot of samples in your val dataset, it should be treated as a 'gallery' style dataset where it's a hand picked limited set of test samples as it does a full similarity matrix. That could really spike memory. We usually use zero-shot eval to gauge progress as it's more sane.

A batch-wise (average over batched similarities) would be possible but not implemented.

**Comment by jn2clark** [comment]
Using CSV datasets with the native implementation will lead to an increase in memory. As @miguelalba96 linked, it is not a bug but expected behavior. The solution is either 
1) use a different dataset format like webdataset. This will be streaming though and not map style.
2) if you want map style with minimal changes then you can move to another structure that has no copy-on-read. Pyarrow has worked in the past.

**Comment by Johnson-yue** [comment]
This bug still happen, I also use csv-dataset, train open_clip L14, with default configuration, but the memory increasing . How to fix it ?  replacing webdataset format ??
